libflux/src/semantic/types.rs

```diff
@@ -116,27 +209,490 @@ impl fmt::Display for MonoType {
     }
 }
 
+impl Substitutable for MonoType {
+    fn apply(self, sub: &Subst) -> Self {
+        match self {
+            MonoType::Bool(t) => MonoType::Bool(t.apply(sub)),
+            MonoType::Int(t) => MonoType::Int(t.apply(sub)),
+            MonoType::Uint(t) => MonoType::Uint(t.apply(sub)),
+            MonoType::Float(t) => MonoType::Float(t.apply(sub)),
+            MonoType::String(t) => MonoType::String(t.apply(sub)),
+            MonoType::Duration(t) => MonoType::Duration(t.apply(sub)),
+            MonoType::Time(t) => MonoType::Time(t.apply(sub)),
+            MonoType::Regexp(t) => MonoType::Regexp(t.apply(sub)),
+            MonoType::Var(tvr) => tvr.apply(sub),
+            MonoType::Arr(arr) => MonoType::Arr(Box::new(arr.apply(sub))),
+            MonoType::Row(obj) => MonoType::Row(Box::new(obj.apply(sub))),
+            MonoType::Fun(fun) => MonoType::Fun(Box::new(fun.apply(sub))),
+        }
+    }
+}
+
+impl Unifiable for MonoType {
+    fn unify(self, with: Self, cons: Const) -> Result {
+        match self {
+            MonoType::Bool(t) => {
+                if let MonoType::Var(tv) = with {
+                    tv.unify(self, cons)
+                } else if let MonoType::Bool(s) = with {
```

Comment:
My reading of this code says that the:

```
if let MonoType::Var(tv) = with {
                    tv.unify(self, cons)
}
```

code can be moved out of the `match self` block. Maybe this is because of the `Self` reference in the `unifiable` trait again?

Reply:
Yep, I agree.

@jlapacik I also think that two nested `match` expression would make the code more readable.